src/neo/Ledger/MemoryPool.cs

```diff
@@ -258,21 +258,25 @@ internal bool CanTransactionFitInPool(Transaction tx)
         /// Note: This must only be called from a single thread (the Blockchain actor). To add a transaction to the pool
         ///       tell the Blockchain actor about the transaction.
         /// </summary>
+        /// <param name="snapshot">Snapshot</param>
         /// <param name="hash"></param>
         /// <param name="tx"></param>
         /// <returns></returns>
-        internal bool TryAdd(UInt256 hash, Transaction tx)
+        internal bool TryAdd(SnapshotView snapshot, UInt256 hash, Transaction tx)
         {
             var poolItem = new PoolItem(tx);
 
+            _txRwLock.EnterWriteLock();
             if (_unsortedTransactions.ContainsKey(hash)) return false;
 
             List<Transaction> removedTransactions = null;
-            _txRwLock.EnterWriteLock();
             try
             {
+                // Check balance
+                BigInteger balance = NativeContract.GAS.BalanceOf(snapshot, tx.Sender);
```

Comment:
In this sense, the key error was to not consider the current Balance of the sender, which is updated in parallel during OnPersist event?

Reply:
I think so.

Reply:
I think we should move the "balance check" to  `BlockChain.OnParallelVerified`, which is right place to verify the transaction.